<a href="https://colab.research.google.com/github/hanlintao/langchain/blob/main/Langchain_02.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**本教程由北京语言大学高级翻译学院教师韩林涛撰写，目的是帮助翻译专业的师生了解如何使用Langchain来解决翻译相关的的问题。**

**第一步：准备工作**

请学习本讲前先学习第一讲：[Langchain_01](https://github.com/hanlintao/langchain/blob/main/langchain_01.ipynb)

In [1]:
!pip install langchain==0.0.343 # 安装LangChain

!pip install openai==0.28 # 安装OpenAI

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.9/1.9 MB 26.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 188.2/188.2 kB 16.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 54.0/54.0 kB 6.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.4/49.4 kB 6.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 76.5/76.5 kB 2.2 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
llmx 0.0.15a0 requires cohere, which is not installed.
llmx 0.0.15a0 requires tiktoken, which is not installed.


In [2]:
#设置OpenAI API Key

import os

os.environ["OPENAI_API_KEY"] = ""

**第二步：安装pypdf并读取PDF文件**

pypdf是一个第三方库，用于处理PDF文件

In [3]:
!pip install pypdf

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 284.0/284.0 kB 6.7 MB/s eta 0:00:00


In [6]:
# 使用pypdf读取一个从网页上下载的PDF文件

from langchain.document_loaders import PyPDFLoader
import os

# 正确的 wget 命令用于下载文件
# 注意：GitHub 上的 PDF 文件通常需要使用 "Raw" URL
pdf_url = "https://raw.githubusercontent.com/hehonghui/awesome-english-ebooks/master/01_economist/te_2024.01.27/TheEconomist.2024.01.27.pdf"
pdf_file = "TheEconomist.2024.01.27.pdf"

# 使用 wget 下载文件
!wget {pdf_url} -O {pdf_file}

# 检查文件是否成功下载
if os.path.exists(pdf_file):
    loader = PyPDFLoader(pdf_file)
    pages = loader.load_and_split()
    print(f"已成功加载 {len(pages)} 页。")
else:
    print("文件下载失败。")


--2024-02-01 10:48:35--  https://raw.githubusercontent.com/hehonghui/awesome-english-ebooks/master/01_economist/te_2024.01.27/TheEconomist.2024.01.27.pdf
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 19081311 (18M) [application/octet-stream]
Saving to: ‘TheEconomist.2024.01.27.pdf’

TheEconomist.2024.0 100%[===================>]  18.20M  --.-KB/s    in 0.06s   

2024-02-01 10:48:36 (301 MB/s) - ‘TheEconomist.2024.01.27.pdf’ saved [19081311/19081311]

已成功加载 323 页。


**备注**

以上使用的PDF文件来自：

https://github.com/hehonghui/awesome-english-ebooks

简介：经济学人(含音频)、纽约客、卫报、连线、大西洋月刊等英语杂志免费下载,支持epub、mobi、pdf格式, 每周更新

In [7]:
# 使用pypdf读取本地文件

# 首先将PDF文件上传至当前Jupyter Notebook对应的文件夹，或者加在Google Drive

# 使用pypdf读取一个从网页上下载的PDF文件

from langchain.document_loaders import PyPDFLoader
import os

pdf_file = "TheEconomist.2024.01.27.pdf"

# 检查文件是否成功下载
if os.path.exists(pdf_file):
    loader = PyPDFLoader(pdf_file)
    pages = loader.load_and_split()
    print(f"已成功加载 {len(pages)} 页。")
else:
    print("文件下载失败。")


已成功加载 323 页。


In [11]:
# 使用LangChain的text_splitter方法对PDF文件进行切割，切割时按照chunk_size来切分

from langchain.text_splitter import RecursiveCharacterTextSplitter

pdf_file = "TheEconomist.2024.01.27.pdf"

loader = PyPDFLoader(pdf_file)

documents = loader.load_and_split()  # 加载并切分 PDF 文件

text_splitter = RecursiveCharacterTextSplitter(chunk_size=1000, chunk_overlap=0)

split_pdf_num = text_splitter.split_documents(documents)

print(f'你的PDF文件被切割成了 {len(split_pdf_num)} 份')

你的PDF文件被切割成了 672 份


In [12]:
# 将chunk_size设置为2000

text_splitter = RecursiveCharacterTextSplitter(chunk_size=2000, chunk_overlap=0)

split_pdf_num = text_splitter.split_documents(documents)

print(f'你的PDF文件被切割成了 {len(split_pdf_num)} 份')

你的PDF文件被切割成了 426 份


In [14]:
# 将chunk_size设置为3000

text_splitter = RecursiveCharacterTextSplitter(chunk_size=3000, chunk_overlap=0)

split_pdf_num = text_splitter.split_documents(documents)

print(f'你的PDF文件被切割成了 {len(split_pdf_num)} 份')

你的PDF文件被切割成了 323 份


通过设置chunk_size的值，将其设置为一个比较大的值，可以确保LangChain能够一页页处理PDF文件。

当chunk_size的值为1000时，整个323页的PDF被切分为672份；值为2000时，PDF被切分为426份；值为3000时，PDF被切分323份。

In [18]:
# 读取第一页

from langchain.document_loaders import PyPDFLoader
from langchain.text_splitter import RecursiveCharacterTextSplitter

pdf_file = "TheEconomist.2024.01.27.pdf"

# 创建 PDF 加载器
loader = PyPDFLoader(pdf_file)

# 加载 PDF 文件的第一页
first_page = loader.load()[0]

# 打印第一页的内容，检查是否正确加载
print("第一页原始内容：")

print(first_page.page_content)  # 假设文档对象有 page_content 属性

# 创建文本切割器
text_splitter = RecursiveCharacterTextSplitter(chunk_size=3000, chunk_overlap=0)

# 将第一页的文档对象放入列表中
documents = [first_page]

# 切割第一页的文档
split_pdf = text_splitter.split_documents(documents)

print(f'PDF文件的第一页被切割成了 {len(split_pdf)} 份')

# 打印切割后的内容
print("\n切割后的内容：")
for doc in split_pdf:
    print(doc.page_content)  # 假设文档对象有 page_content 属性


第一页原始内容：

PDF文件的第一页被切割成了 0 份

切割后的内容：


In [31]:
# 由于该PDF的第一页是图片，所以没有任何内容，改为读取第二页

from langchain.document_loaders import PyPDFLoader
from langchain.text_splitter import RecursiveCharacterTextSplitter

pdf_file = "TheEconomist.2024.01.27.pdf"

# 创建 PDF 加载器
loader = PyPDFLoader(pdf_file)

# 加载 PDF 文件的第二页
first_page = loader.load()[1]

# 打印第二页的内容，检查是否正确加载
print("第二页原始内容：")

print(first_page.page_content)  # 假设文档对象有 page_content 属性

# 创建文本切割器
text_splitter = RecursiveCharacterTextSplitter(chunk_size=3000, chunk_overlap=0)

# 将第二页的文档对象放入列表中
documents = [first_page]

# 切割第二页的文档
split_pdf = text_splitter.split_documents(documents)

print(f'PDF文件的第二页被切割成了 {len(split_pdf)} 份')

# 打印切割后的内容
print("\n切割后的内容：")
for doc in split_pdf:
    print(doc.page_content)  # 假设文档对象有 page_content 属性


第二页原始内容：
 
The world this week
Leaders
Letters
By Invitation
Briefing
Asia
China
United States
Middle East & Africa
The Americas
Europe
Britain
International
Business
Finance & economics
Science & technology
Cultur e
The Economist r eads
Economic & financial indicators
Obituary
 
PDF文件的第二页被切割成了 1 份

切割后的内容：
The world this week
Leaders
Letters
By Invitation
Briefing
Asia
China
United States
Middle East & Africa
The Americas
Europe
Britain
International
Business
Finance & economics
Science & technology
Cultur e
The Economist r eads
Economic & financial indicators
Obituary


In [28]:
# 读取PDF文件特定范围的内容

# 当前PDF显示：从第21页24页有一篇文章，名为：AI holds tantalising promise for the emerging world

from langchain.document_loaders import PyPDFLoader
from langchain.text_splitter import RecursiveCharacterTextSplitter

pdf_file = "TheEconomist.2024.01.27.pdf"

# 创建 PDF 加载器
loader = PyPDFLoader(pdf_file)

# 加载 PDF 文件的第 21 到第 24 页
# 注意：页码索引从 0 开始，因此第 21 页是索引 20，第 24 页是索引 23
pages_21_to_24 = loader.load()[20:24]

# 遍历每一页并打印其内容
for i, page in enumerate(pages_21_to_24, start=21):  # 从第 21 页开始计数
    print(f"第 {i} 页内容：")
    print(page.page_content)  # 假设文档对象有 page_content 属性
    print("\n" + "-"*50 + "\n")

# 创建文本切割器
text_splitter = RecursiveCharacterTextSplitter(chunk_size=3000, chunk_overlap=0)


# 切割第 21 到第 24 页的文档
split_pdf = text_splitter.split_documents(pages_21_to_24)

print(f'PDF文件的第21到24页被切割成了 {len(split_pdf)} 份')

# 打印切割后的内容
print("\n切割后的内容：\n")
for doc in split_pdf:
    print( doc.page_content + "\n" + "-"*50 + "\n")  # 假设文档对象有 page_content 属性



第 21 页内容：
AI for all
AI holds tantalising promise for the emerging
world
It could help boost human capital, and ultimately growth
Jan 25th 2024
NEW TECHNO LOGY brings with it both the sweet hope of greater
prosperity and the cruel fear of missing out. Satya Nadella, the boss of
Microsoft, says he is haunted by the fact that the Industrial Revolution left
behind India, his country of birth. (Indian manufacturers hardly enjoyed a
level playing-fie ld—Britain was then both their rival and their ruler.) Many
technologies, such as online-education courses, have generated more hype
than economic growth in the emer ging world. Some people  worry that
generative artificial intelligence  (AI), too, will disappoint the global south.
The big winners so far seem to be a bunch of Western early adop ters, as well
as startups in San Francisco and America’ s “magnificent seven ” tech firms,
which include Microsoft and have together added an astonishing $4.6trn to
their market value since ChatGPT’ s la

In [29]:
# 通过观察上面的内容发现：PDF换页后会把文本内容切分，造成不完整的句子，此时我们来看overlap的值是否能解决问题

# 当前PDF显示：从第21页24页有一篇文章，名为：AI holds tantalising promise for the emerging world

from langchain.document_loaders import PyPDFLoader
from langchain.text_splitter import RecursiveCharacterTextSplitter

pdf_file = "TheEconomist.2024.01.27.pdf"

# 创建 PDF 加载器
loader = PyPDFLoader(pdf_file)

# 加载 PDF 文件的第 21 到第 24 页
# 注意：页码索引从 0 开始，因此第 21 页是索引 20，第 24 页是索引 23
pages_21_to_24 = loader.load()[20:24]

# 遍历每一页并打印其内容
for i, page in enumerate(pages_21_to_24, start=21):  # 从第 21 页开始计数
    print(f"第 {i} 页内容：")
    print(page.page_content)  # 假设文档对象有 page_content 属性
    print("\n" + "-"*50 + "\n")

# 创建文本切割器
text_splitter = RecursiveCharacterTextSplitter(chunk_size=2000, chunk_overlap=100)


# 切割第 21 到第 24 页的文档
split_pdf = text_splitter.split_documents(pages_21_to_24)

print(f'PDF文件的第21到24页被切割成了 {len(split_pdf)} 份')

# 打印切割后的内容
print("\n切割后的内容：\n")
for doc in split_pdf:
    print( doc.page_content + "\n" + "-"*50 + "\n")  # 假设文档对象有 page_content 属性




第 21 页内容：
AI for all
AI holds tantalising promise for the emerging
world
It could help boost human capital, and ultimately growth
Jan 25th 2024
NEW TECHNO LOGY brings with it both the sweet hope of greater
prosperity and the cruel fear of missing out. Satya Nadella, the boss of
Microsoft, says he is haunted by the fact that the Industrial Revolution left
behind India, his country of birth. (Indian manufacturers hardly enjoyed a
level playing-fie ld—Britain was then both their rival and their ruler.) Many
technologies, such as online-education courses, have generated more hype
than economic growth in the emer ging world. Some people  worry that
generative artificial intelligence  (AI), too, will disappoint the global south.
The big winners so far seem to be a bunch of Western early adop ters, as well
as startups in San Francisco and America’ s “magnificent seven ” tech firms,
which include Microsoft and have together added an astonishing $4.6trn to
their market value since ChatGPT’ s la

通过修改chunk_size和overlap的值发现：PDF本身依然是安装页进行了切分，切分之后每一页内部按照chunk_size进一步切分，而overlap仅在这个过程中起到作用，并没有原始的PDF文件跨页分割的内容合并到一起。

上面的例子能够帮助我们直观理解一个PDF文件是如何被切分成页，以及页又如何被切分成chunk。

**第三步：生成PDF文件中特定文章的摘要**

方法一：直接生成整个文章的摘要

In [30]:
# 通过观察上面的内容发现：PDF换页后会把文本内容切分，造成不完整的句子，此时我们来看overlap的值是否能解决问题

# 当前PDF显示：从第21页24页有一篇文章，名为：AI holds tantalising promise for the emerging world

# 由于该PDF的第一页是图片，所以没有任何内容，改为读取第二页

from langchain.document_loaders import PyPDFLoader
from langchain.text_splitter import RecursiveCharacterTextSplitter

pdf_file = "TheEconomist.2024.01.27.pdf"

# 创建 PDF 加载器
loader = PyPDFLoader(pdf_file)

# 加载 PDF 文件的第 21 到第 24 页
# 注意：页码索引从 0 开始，因此第 21 页是索引 20，第 24 页是索引 23
pages_21_to_24 = loader.load()[20:24]

whole_content = ""

# 遍历每一页并打印其内容
for i, page in enumerate(pages_21_to_24, start=21):  # 从第 21 页开始计数
    #print(f"第 {i} 页内容：")
    #print(page.page_content)  # 假设文档对象有 page_content 属性
    #print("\n" + "-"*50 + "\n")

    whole_content += page.page_content

print(f"第 21 到第 24 页页内容：\n")

print(whole_content)


第 21 到第 24 页页内容：

AI for all
AI holds tantalising promise for the emerging
world
It could help boost human capital, and ultimately growth
Jan 25th 2024
NEW TECHNO LOGY brings with it both the sweet hope of greater
prosperity and the cruel fear of missing out. Satya Nadella, the boss of
Microsoft, says he is haunted by the fact that the Industrial Revolution left
behind India, his country of birth. (Indian manufacturers hardly enjoyed a
level playing-fie ld—Britain was then both their rival and their ruler.) Many
technologies, such as online-education courses, have generated more hype
than economic growth in the emer ging world. Some people  worry that
generative artificial intelligence  (AI), too, will disappoint the global south.
The big winners so far seem to be a bunch of Western early adop ters, as well
as startups in San Francisco and America’ s “magnificent seven ” tech firms,
which include Microsoft and have together added an astonishing $4.6trn to
their market value since ChatG

In [32]:
# 使用LangChain来对整个内容进行摘要

# 设计提问模板

from langchain.prompts.chat import (
    SystemMessagePromptTemplate,
    HumanMessagePromptTemplate,
    ChatPromptTemplate
)

system_template = SystemMessagePromptTemplate.from_template("你是北京语言大学高级翻译学院教授高级英语的教师，在给学生讲解经济学人之前会先对文章的核心内容进行分析")

user_template = HumanMessagePromptTemplate.from_template("{user_prompt}")

template = ChatPromptTemplate.from_messages([system_template, user_template])

# 创建OpenAI问答实例

from langchain.chat_models import ChatOpenAI

llm = ChatOpenAI(temperature=0.7, model_name="gpt-4")

# 创建一个语言链

from langchain.chains import LLMChain

chain = LLMChain(llm=llm, prompt=template)

# 创建用户问题

user_prompt = "请总结以下经济学人文章的核心观点和论据：" + whole_content

# 运行语言链

chain_output = chain.run({"user_prompt": user_prompt})

print(chain_output)


这篇文章的核心观点是，人工智能（AI）对发展中国家具有巨大的潜在价值，可以提高生产力，减少人力资本的差距，改善公共服务的可及性和质量，从而有助于推动经济增长。

论据包括：

1. AI技术的推广可以使发展中国家的生产力提高，人力资本差距缩小。比如，印度正在使用大型语言模型和语音识别软件，使文盲农民能够询问机器人如何申请政府贷款；肯尼亚的学生们将能够向聊天机器人提问他们的作业问题，而聊天机器人将根据反馈调整和改进教学方法。

2. AI可以帮助提高公共服务的质量和可及性，从而有助于发展中国家解决教育和健康的问题。例如，巴西的研究人员正在测试一种医疗AI，该AI可以帮助基层医疗工作者诊疗病人。

3. AI可以通过手机等设备快速传播，比20世纪初的新技术传播得更快。此外，AI技术可以根据当地的需求进行定制，不同的方法都有可能成功。

4. 一些国家已经开始利用AI技术。例如，中国的技术实力仅次于美国，得益于其科技知识和互联网巨头的深厚财力；印度虽然可能因AI的发展受到影响，但其充满活力的创业环境，众多的技术开发者以及渴望利用AI改善数字基础设施的政府，使其具备创新和适应的能力。

5. 但是，AI的发展也面临挑战，比如计算能力可能变得过于昂贵，需要收集和存储的数据量巨大，一些从业者可能无法充分利用手中的知识，或者没有尝试新事物的动力。

总的来说，尽管AI的发展还存在许多不确定性，但对于发展中国家来说，AI无疑是一个巨大的机遇，并且他们有能力去抓住这个机遇。


In [33]:
# 引入价格计算和实践计算

from langchain import PromptTemplate  # 引入模板
from langchain.chat_models import ChatOpenAI  # 引入模型
from langchain.chains import LLMChain  # 引入语言链
from langchain.callbacks import get_openai_callback # 引入OpenAI Callback，用于计算费用
import time #用于计算耗时

# 创建模板
system_template = "你是北京语言大学高级翻译学院教授高级英语的教师，在给学生讲解经济学人之前会先对文章的核心内容进行分析"
user_template = "请总结以下经济学人文章的核心观点和论据：：{sentences}"

template = PromptTemplate(input_variables=["sentences"], template=user_template)

# 创建语言模型
llm = ChatOpenAI(temperature=0.7, model_name="gpt-4")

# 创建语言链
chain = LLMChain(llm=llm, prompt=template)


total_cost = 0  # 初始化总费用
total_time = 0  # 初始化总耗时

# 生成输入
user_prompt = {"sentences": whole_content}

# 记录开始时间
start_time = time.time()

# 使用OpenAI Callback计算费用
with get_openai_callback() as cb:
    # 运行语言链
    chain_output = chain.run(user_prompt)

    # 本次调用的费用
    cost = cb.total_cost
    total_cost = cost  # 更新总费用

    # 计算完成这一行的耗时
    row_time = time.time() - start_time
    total_time = row_time  # 更新总耗时

    # 输出翻译结果
    print(f"摘要: {chain_output}\n耗时：{total_time}秒\n耗费：{total_cost}美元\n")



摘要: 核心观点：
这篇文章主要讨论了人工智能（AI）对发展中国家的潜在影响和好处。作者认为，AI具有提高生产力和缩小人力资本差距的潜力，能够帮助发展中国家提高收入水平，提供更好和更易获取的公共服务。尽管之前的一些技术并未在发展中国家带来经济增长，但AI有可能改变这一现状。AI的传播速度也比20世纪早期的技术快得多，且可以根据当地需求进行调整。此外，一些国家如中国、印度和海湾国家已经开始利用AI来改善其数字基础设施。

论据：
文章列举了一些具体的例子来支持其观点。例如，印度正在结合大型语言模型和语音识别软件，使文盲农民能够询问如何申请政府贷款；肯尼亚的学生将通过聊天机器人提问作业问题，而聊天机器人将根据反馈调整和改进其教学内容；巴西的研究人员正在测试一种可以帮助初级医疗工作者治疗病人的医疗AI。此外，文章还提到，AI的传播将通过手机这种已经广泛使用的设备来实现，这将加速其在发展中国家的普及。
耗时：29.614582300186157秒
耗费：0.07083美元



In [34]:
# 切换成更便宜的模型

# 引入价格计算和实践计算

from langchain import PromptTemplate  # 引入模板
from langchain.chat_models import ChatOpenAI  # 引入模型
from langchain.chains import LLMChain  # 引入语言链
from langchain.callbacks import get_openai_callback # 引入OpenAI Callback，用于计算费用
import time #用于计算耗时

# 创建模板
system_template = "你是北京语言大学高级翻译学院教授高级英语的教师，在给学生讲解经济学人之前会先对文章的核心内容进行分析"
user_template = "请总结以下经济学人文章的核心观点和论据：：{sentences}"

template = PromptTemplate(input_variables=["sentences"], template=user_template)

# 创建语言模型
llm = ChatOpenAI(temperature=0.7, model_name="gpt-3.5-turbo-1106")

# 创建语言链
chain = LLMChain(llm=llm, prompt=template)


total_cost = 0  # 初始化总费用
total_time = 0  # 初始化总耗时

# 生成输入
user_prompt = {"sentences": whole_content}

# 记录开始时间
start_time = time.time()

# 使用OpenAI Callback计算费用
with get_openai_callback() as cb:
    # 运行语言链
    chain_output = chain.run(user_prompt)

    # 本次调用的费用
    cost = cb.total_cost
    total_cost = cost  # 更新总费用

    # 计算完成这一行的耗时
    row_time = time.time() - start_time
    total_time = row_time  # 更新总耗时

    # 输出翻译结果
    print(f"摘要: {chain_output}\n耗时：{total_time}秒\n耗费：{total_cost}美元\n")



摘要: 核心观点：
人工智能（AI）在新兴世界中具有令人向往的潜力，可以帮助提高人力资本，最终促进经济增长。

论据：
1. AI在新兴世界中有可能提高生产率，缩小人力资本差距，帮助收入水平赶上富裕世界。
2. AI将成为消费者和工人的有用多功能工具，使获取和解释信息更容易，创造新的工作岗位。
3. AI可以改善和提供更易获得的公共服务，解决新兴经济体长期以来因教育和健康工作者匮乏而受阻的问题。
4. AI技术将通过新兴世界已经普及的手机迅速传播，且可以根据当地需求进行定制。
5. 许多国家已经在利用AI，如中国、印度和海湾国家，它们都有能力创新和适应技术。
6. 尽管AI技术仍在不断发展，但是对于加速AI传播的投资将得到丰厚的回报。

总结：
人工智能在新兴世界中具有巨大的潜力，可以帮助提高人力资本，改善公共服务，并加速经济增长。尽管存在一些挑战，但是新兴国家有机会抓住这一机遇，实现经济发展。
耗时：12.220199346542358秒
耗费：0.002371美元



通过对比发现，使用gpt-4模型，费用为：0.07083美元；使用gpt-3.5-turbo-1106模型，费用为0.002371美元。

根据OpenAI的价格介绍：https://openai.com/pricing

现在最贵的模型是GPT-4系列模型（包括：gpt-4-32k和gpt-4），其次是GPT-4 Turbo系列模型（包括：gpt-4-0125-preview、gpt-4-1106-preview、gpt-4-1106-vision-preview），最便宜的是GPT-3.5 Turbo系列模型（包括：gpt-3.5-turbo-1106、gpt-3.5-turbo-instruct）

方法二：先生成每一个chunk的摘要，再合并成一个新的摘要

In [39]:
# 切换成GPT-4-turbo

from langchain.document_loaders import PyPDFLoader
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain import PromptTemplate
from langchain.chat_models import ChatOpenAI
from langchain.chains import LLMChain
from langchain.callbacks import get_openai_callback
import time

# 创建 PDF 加载器和文本切割器
pdf_file = "TheEconomist.2024.01.27.pdf"
loader = PyPDFLoader(pdf_file)
text_splitter = RecursiveCharacterTextSplitter(chunk_size=2000, chunk_overlap=100)
pages_21_to_24 = loader.load()[20:24]
split_pdf = text_splitter.split_documents(pages_21_to_24)

# 创建语言模型和链
llm = ChatOpenAI(temperature=0.7, model_name="gpt-4-1106-preview")
chain = LLMChain(llm=llm, prompt=PromptTemplate(input_variables=["sentences"], template="请总结以下经济学人文章的核心观点和论据：{sentences}"))

# 为每个 chunk 生成摘要
individual_summaries = []
total_cost = 0
total_time = 0

for doc in split_pdf:
    start_time = time.time()
    with get_openai_callback() as cb:
        summary = chain.run({"sentences": doc.page_content})
        individual_summaries.append(summary)
        cost = cb.total_cost
        total_cost += cost
        row_time = time.time() - start_time
    total_time += row_time
    # 输出摘要结果
    print(f"摘要: {summary}\n耗时：{row_time}秒\n耗费：{cost}美元\n")

# 将所有摘要合并
combined_summary = ' '.join(individual_summaries)

# 为合并的摘要生成新摘要
start_time = time.time()
with get_openai_callback() as cb:
    final_summary = chain.run({"sentences": combined_summary})
    total_cost += cb.total_cost
    total_time_final = time.time() - start_time
    print(f"合并摘要: {final_summary}\n耗时：{total_time_final}秒\n耗费：{cost}美元\n")
total_time += total_time_final

# 输出结果
print(f"单独摘要：\n{' '.join(individual_summaries)}")
print(f"\n综合摘要：\n{final_summary}")
print(f"\n总耗时：{total_time}秒\n总费用：{total_cost}美元")


摘要: 核心观点：
文章提出了人工智能（AI）对新兴世界带来的希望和潜力，同时也表达了对于技术可能不平等分布的担忧。文章指出，尽管目前AI技术的主要受益者似乎是西方早期采纳者以及硅谷的初创公司和美国七大科技巨头（包括微软），但AI有可能在新兴国家和地区也产生积极影响。通过提高生产力和缩小人力资本差距，AI有望促进这些国家的经济增长。

论据：
1. 微软CEO萨提亚·纳德拉对工业革命时期印度的遗漏表达了遗憾，这反映了对过去技术变革中不平等分布的担忧。
2. 过去，一些技术（如在线教育课程）在新兴世界产生的热潮并未能转化为实质性的经济增长。
3. 自从ChatGPT在2022年11月推出以来，美国七大科技公司市值惊人地增加了4.6万亿美元，这突显了AI技术在一些地区和公司中的快速采纳和巨大影响。
4. 文章暗示，AI技术的传播有潜力在新兴国家和地区带来转型，通过提高生产力和改善人力资本来促进增长。
耗时：22.050694942474365秒
耗费：0.016309999999999998美元

摘要: 核心观点:
文章指出人工智能（AI）在发展中国家的应用具有极大的潜力，不仅能成为消费者和工人的有用工具，还能通过改善公共服务质量来推动社会经济转型。与富裕国家相比，发展中国家的白领工作岗位较少，因此AI带来的颠覆性变化可能较小，但新工作岗位的创造也可能较少。国际货币基金组织（IMF）预测，发展中国家有五分之一到四分之一的劳动力面临被AI替代的风险，而在富裕国家这一比例为三分之一。

论据:
1. 发展中国家长期以来受到教育水平低下和医疗资源匮乏的制约，导致健康状况和生产力不足。
2. AI的应用可以通过提供更好的教育和医疗服务来克服这些障碍。例如，印度正在开发结合大型语言模型和语音识别软件的系统，使不识字的农民能够向机器人询问如何申请政府贷款；肯尼亚的学生将能够向聊天机器人提问作业问题，而机器人则能够根据反馈调整并改进其教学内容。
3. 在巴西，研究人员正在测试一种医疗AI，帮助训练不足的基层医疗工作者治疗病人。全球收集的医疗数据被输入到AI系统中，可以帮助提高诊断的准确性。

总结来说，文章认为AI技术能够支持发展中国家在教育和医疗领域取得突破，从而帮助这些国家提升人力资本质量，减少与富裕国家之间的收入差距。发展中国家不应该只是被动接受AI，而应该主动塑造AI以满足

**第四步：使用LangChain直接对PDF文件进行总结**

In [55]:
# 当前PDF显示：从第21页24页有一篇文章，名为：AI holds tantalising promise for the emerging world

from langchain import PromptTemplate # 引入模板
from langchain.document_loaders import PyPDFLoader
from langchain.chat_models import ChatOpenAI
from langchain.chains.summarize import load_summarize_chain
from langchain.callbacks import get_openai_callback # 引入OpenAI Callback，用于计算费用
import time #用于计算耗时

pdf_file = "TheEconomist.2024.01.27.pdf"

# 创建 PDF 加载器
loader = PyPDFLoader(pdf_file)

# 加载 PDF 文件的第 21 到第 24 页
# 注意：页码索引从 0 开始，因此第 21 页是索引 20，第 24 页是索引 23
pages_21_to_24 = loader.load()[20:24]

# 创建模板
system_template = "你是北京语言大学高级翻译学院教授高级英语的教师，在给学生讲解经济学人之前会先对文章的核心内容进行分析"
user_template = "请总结以下经济学人文章的核心观点和论据：{text}"

template = PromptTemplate(input_variables=["text"], template=user_template)

# 创建语言模型
llm = ChatOpenAI(temperature=0.7, model_name="gpt-3.5-turbo-1106")

# 创建链

chain = load_summarize_chain(llm,chain_type="stuff",prompt=template)

total_cost = 0  # 初始化总费用
total_time = 0  # 初始化总耗时

# 记录开始时间
start_time = time.time()

# 使用OpenAI Callback计算费用
with get_openai_callback() as cb:
    # 运行语言链
    chain_output = chain.run(pages_21_to_24)

    # 本次调用的费用
    cost = cb.total_cost
    total_cost = cost  # 更新总费用

    # 计算完成这一行的耗时
    row_time = time.time() - start_time
    total_time = row_time  # 更新总耗时

    # 输出翻译结果
    print(f"摘要: {chain_output}\n耗时：{total_time}秒\n耗费：{total_cost}美元\n")

摘要: 核心观点：人工智能在发展中国家具有潜在的巨大利益，可以提高人力资本和最终推动经济增长。发展中国家并不需要被动接受人工智能，而是可以塑造它以满足自己的需求。

论据：
1. 人工智能有可能提高生产率，缩小人力资本差距，并帮助发展中国家的收入水平赶上富裕国家。
2. 发展中国家可以利用人工智能改善公共服务，解决教育和医疗领域的问题。
3. 人工智能技术已经在一些发展中国家得到应用，如印度结合大型语言模型和语音识别软件帮助文盲农民申请政府贷款。
4. 人工智能技术的传播速度将比早期的技术快得多，因为它可以通过智能手机迅速传播。
5. 不同国家可以根据自己的需求定制人工智能技术，而不受美国社交媒体和互联网搜索公司的垄断影响。
6. 尽管人工智能的发展还存在许多不确定性，但发展中国家有机会利用这一技术，而且投资促进人工智能传播的回报将丰厚。
耗时：20.867209434509277秒
耗费：0.002333美元



In [56]:
#切换为GPT-4 Turbo
# 当前PDF显示：从第21页24页有一篇文章，名为：AI holds tantalising promise for the emerging world

from langchain import PromptTemplate # 引入模板
from langchain.document_loaders import PyPDFLoader
from langchain.chat_models import ChatOpenAI
from langchain.chains.summarize import load_summarize_chain
from langchain.callbacks import get_openai_callback # 引入OpenAI Callback，用于计算费用
import time #用于计算耗时

pdf_file = "TheEconomist.2024.01.27.pdf"

# 创建 PDF 加载器
loader = PyPDFLoader(pdf_file)

# 加载 PDF 文件的第 21 到第 24 页
# 注意：页码索引从 0 开始，因此第 21 页是索引 20，第 24 页是索引 23
pages_21_to_24 = loader.load()[20:24]

# 创建模板
system_template = "你是北京语言大学高级翻译学院教授高级英语的教师，在给学生讲解经济学人之前会先对文章的核心内容进行分析"
user_template = "请总结以下经济学人文章的核心观点和论据：{text}"

template = PromptTemplate(input_variables=["text"], template=user_template)

# 创建语言模型
llm = ChatOpenAI(temperature=0.7, model_name="gpt-4-1106-preview")

# 创建链

chain = load_summarize_chain(llm,chain_type="stuff",prompt=template)

total_cost = 0  # 初始化总费用
total_time = 0  # 初始化总耗时

# 记录开始时间
start_time = time.time()

# 使用OpenAI Callback计算费用
with get_openai_callback() as cb:
    # 运行语言链
    chain_output = chain.run(pages_21_to_24)

    # 本次调用的费用
    cost = cb.total_cost
    total_cost = cost  # 更新总费用

    # 计算完成这一行的耗时
    row_time = time.time() - start_time
    total_time = row_time  # 更新总耗时

    # 输出翻译结果
    print(f"摘要: {chain_output}\n耗时：{total_time}秒\n耗费：{total_cost}美元\n")

摘要: 核心观点：
文章指出，人工智能（AI）对发展中国家来说充满了巨大的潜力。尽管人们担心AI可能会让全球南部地区失望，但文章认为AI有可能提高生产力、缩小人力资本差距，并最终帮助这些国家实现经济增长，从而赶上富裕国家的水平。

论据：
1. AI作为一个全能的工具，将使消费者和工人更容易获取和解释信息，尽管一些工作可能会消失，但也会创造新的工作机会。
2. 因为发展中国家的白领工人较少，所以与西方相比，AI引起的颠覆可能更小，但对现有企业的增益可能也更小。
3. 发展中国家长期以来受到教育水平低和健康工作者缺乏的制约，AI有可能通过改善公共服务质量来改变这一局面，例如，通过AI提高教育和医疗服务的可用性和质量。
4. 与20世纪初的技术相比，AI的传播速度可能更快，因为许多发展中国家的人们已经拥有了手机，这是AI技术传播的载体。
5. AI技术可以根据当地需要进行定制，例如印度开发人员正在用当地数据微调西方模型以提供翻译服务，而不需要承担建模的高昂资本成本。
6. 一些国家（如中国和印度）已经在利用AI技术，并且正在寻求通过AI提升其数字基础设施。

尽管文章也提到了一些挑战，比如计算能力可能变得过于昂贵，需要收集和存储本地数据，以及缺乏利用新知识的能力或动机，但总体上，它强调对AI的投资将带来丰厚的回报。发展中国家有机会利用AI实现突破，而且他们有能力抓住这个机会。

下载来源说明：本文由zlibrary从经济学人网站下载，属于订阅者专享内容。
耗时：62.423388719558716秒
耗费：0.03428美元



小结

通过上述内容我们解了如何读取PDF文件，并针对PDF文件进行摘要。我们可以直接读取PDF文件的每一页的内容，拼接在一起后去摘要；也可以把每一页内容单独做完摘要后，将多个摘要合并到一起再做一次摘要；也可以使用LangChain的load_summarize_chain直接对一个PDF文件做摘要。

通过上面的逐步分析，我们发现，最好的办法是直接使用LangChain的摘要功能。

基本逻辑如下：


• 导入必要的LangChain模块和类,包括模板、PDF加载器、语言模型、链等。

• 加载指定路径下的PDF文件,并提取第21-24页的内容。

• 构建一个系统模板和用户模板,分别表示教师背景和总结文章要求。组合成一个PromptTemplate。

• 创建一个ChatOpenAI的语言模型实例,指定温度参数和GPT-3.5模型。

• 使用load_summarize_chain函数创建一个文本摘要的语言链实例,同时传入语言模型实例和自定义模板。

• 初始化总费用和总耗时计数器。

• 使用OpenAI Callback计算调用费用。

• 运行语言链,生成文本摘要,同时计算本次调用耗时和费用。

• 输出摘要结果、总耗时和总费用。

**第五步：使用map_reduce模式和refine模式处理长文本**

我们前面介绍了两种做摘要的模式：直接处理全部文本；分开处理单个文本块儿生成单个摘要，再对所有单个摘要汇总进行一次摘要。

通过分析发现第二种模式似乎效果不佳，所以直接处理全部文本是简单的，而且有了Langchain的summarization功能，似乎代码更简单了。

但是，一旦遇到更长的文本，这个方法就不奏效了。

我们尝试读取100页PDF文件，并使用GPT 3.5 Turbo来对这100页做摘要：

In [59]:


from langchain import PromptTemplate # 引入模板
from langchain.document_loaders import PyPDFLoader
from langchain.chat_models import ChatOpenAI
from langchain.chains.summarize import load_summarize_chain
from langchain.callbacks import get_openai_callback # 引入OpenAI Callback，用于计算费用
import time #用于计算耗时

pdf_file = "TheEconomist.2024.01.27.pdf"

# 创建 PDF 加载器
loader = PyPDFLoader(pdf_file)

# 加载 PDF 文件的第 21 到第 121 页
# 注意：页码索引从 0 开始，因此第 21 页是索引 20，第 121 页是索引 120
pages_21_to_121 = loader.load()[20:120]

# 创建模板
system_template = "你是北京语言大学高级翻译学院教授高级英语的教师，在给学生讲解经济学人之前会先对文章的核心内容进行分析"
user_template = "请总结以下经济学人文章的核心观点和论据：{text}"

template = PromptTemplate(input_variables=["text"], template=user_template)

# 创建语言模型
llm = ChatOpenAI(temperature=0.7, model_name="gpt-3.5-turbo-1106")

# 创建链

chain = load_summarize_chain(llm,chain_type="stuff",prompt=template)

total_cost = 0  # 初始化总费用
total_time = 0  # 初始化总耗时

# 记录开始时间
start_time = time.time()

# 使用OpenAI Callback计算费用
with get_openai_callback() as cb:
    # 运行语言链
    chain_output = chain.run(pages_21_to_121)

    # 本次调用的费用
    cost = cb.total_cost
    total_cost = cost  # 更新总费用

    # 计算完成这一行的耗时
    row_time = time.time() - start_time
    total_time = row_time  # 更新总耗时

    # 输出翻译结果
    print(f"摘要: {chain_output}\n耗时：{total_time}秒\n耗费：{total_cost}美元\n")

InvalidRequestError: This model's maximum context length is 16385 tokens. However, your messages resulted in 31515 tokens. Please reduce the length of the messages.

报错如下：This model's maximum context length is 16385 tokens. However, your messages resulted in 31515 tokens. Please reduce the length of the messages.

也就是说：GPT 3.5 Turbo最多只能接受16385个Token，但是这里的100页文件有31515个Token。

目前只有GPT-4 Turbo支持的Token最大：128000个Token，但是价格也非常昂贵。
from langchain import ChatChain, PromptTemplate
from langchain.document_loaders import PyPDFLoader
from langchain.chat_models import ChatOpenAI
from langchain.chains import MapReduceDocumentsChain, ReduceDocumentsChain
from langchain.callbacks import get_openai_callback
所以，为了节约成本，我们只能选取另一种模式：map_reduce，而之前的模式名为：stuff，如下面的代码所示：

`chain = load_summarize_chain(llm,chain_type="stuff",prompt=template) `


In [69]:
!pip install tiktoken

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 27.8 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
llmx 0.0.15a0 requires cohere, which is not installed.


In [74]:
# 引入mapreduce模式

from langchain import PromptTemplate # 引入模板
from langchain.document_loaders import PyPDFLoader
from langchain.chat_models import ChatOpenAI
from langchain.chains.summarize import load_summarize_chain
from langchain.chains.mapreduce import MapReduceChain # 引入 MapReduceChain用于处理大文件
from langchain.callbacks import get_openai_callback # 引入OpenAI Callback，用于计算费用
import time #用于计算耗时

pdf_file = "TheEconomist.2024.01.27.pdf"

# 创建 PDF 加载器
loader = PyPDFLoader(pdf_file)

# 加载 PDF 文件的第 21 到第 121 页
# 注意：页码索引从 0 开始，因此第 21 页是索引 20，第 121 页是索引 120
pages_21_to_121 = loader.load()[20:120]

# 创建模板
system_template = "你是北京语言大学高级翻译学院教授高级英语的教师，在给学生讲解经济学人之前会先对文章的核心内容进行分析"
user_template = "请总结以下经济学人文章的核心观点和论据：{text}"

template = PromptTemplate(input_variables=["text"], template=user_template)

# 创建语言模型
llm = ChatOpenAI(temperature=0.7, model_name="gpt-3.5-turbo-1106")

# 创建链

chain = load_summarize_chain(llm,chain_type="map_reduce", map_prompt=template, combine_prompt=template)

total_cost = 0  # 初始化总费用
total_time = 0  # 初始化总耗时

# 记录开始时间
start_time = time.time()

# 使用OpenAI Callback计算费用
with get_openai_callback() as cb:
    # 运行语言链
    chain_output = chain.run(pages_21_to_121)

    # 本次调用的费用
    cost = cb.total_cost
    total_cost = cost  # 更新总费用

    # 计算完成这一行的耗时
    row_time = time.time() - start_time
    total_time = row_time  # 更新总耗时

    # 输出翻译结果
    print(f"摘要: {chain_output}\n耗时：{total_time}秒\n耗费：{total_cost}美元\n")

  "error": {
    "message": "The server had an error while processing your request. Sorry about that!",
    "type": "server_error",
    "param": null,
    "code": null
  }
}
 500 {'error': {'message': 'The server had an error while processing your request. Sorry about that!', 'type': 'server_error', 'param': None, 'code': None}} {'Date': 'Thu, 01 Feb 2024 14:38:57 GMT', 'Content-Type': 'application/json', 'Content-Length': '176', 'Connection': 'keep-alive', 'access-control-allow-origin': '*', 'openai-model': 'gpt-3.5-turbo-1106', 'openai-organization': 'user-cuw19g1klu0rkvj8nuisza87', 'openai-processing-ms': '3258', 'openai-version': '2020-10-01', 'strict-transport-security': 'max-age=15724800; includeSubDomains', 'x-ratelimit-limit-requests': '5000', 'x-ratelimit-limit-tokens': '80000', 'x-ratelimit-remaining-requests': '4999', 'x-ratelimit-remaining-tokens': '79487', 'x-ratelimit-reset-requests': '12ms', 'x-ratelimit-reset-tokens': '384ms', 'x-request-id': '723c649dae15451b94dc204b49

摘要: 

核心观点：文章认为梅洛尼并没有成为一个小特朗普，她的政府通过了一些法律旨在帮助小企业，她的温和态度部分是出于计算，也反映了她自身的信仰。

论据：意大利的GDP增长，梅洛尼政府通过了帮助小企业的法律，她并没有试图让意大利退出欧盟或破坏法治，她的温和态度是出于计算和反映了她的信仰。

总的来说，梅洛尼似乎在真诚地努力使她的国家变得更好，而不是让它崩溃。
耗时：812.1583223342896秒
耗费：0.106515美元



在完成上面的100页任务总结过程中，大概在运行了10分钟后，出现下方的报错：

`WARNING:langchain_core.language_models.llms:Retrying langchain.chat_models.openai.ChatOpenAI.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised Timeout: Request timed out: HTTPSConnectionPool(host='api.openai.com', port=443): Read timed out. (read timeout=600).`

在运行20分钟后给出以下回答：

摘要:

核心观点：文章认为梅洛尼并没有成为一个小特朗普，她的政府通过了一些法律旨在帮助小企业，她的温和态度部分是出于计算，也反映了她自身的信仰。

论据：意大利的GDP增长，梅洛尼政府通过了帮助小企业的法律，她并没有试图让意大利退出欧盟或破坏法治，她的温和态度是出于计算和反映了她的信仰。

总的来说，梅洛尼似乎在真诚地努力使她的国家变得更好，而不是让它崩溃。
耗时：812.1583223342896秒
耗费：0.106515美元

考虑到经济学人这100页的内容涵盖的主题非常多，最后的摘要质量并不高。

因为这类任务的耗时时间非常长，所以一般不建议针对太长的PDF文件进行内容总结，map_ruduce模式缺失的信息太多。

接下来引入第三种模式：refine模式

In [75]:
# 引入refine模式

from langchain import PromptTemplate # 引入模板
from langchain.document_loaders import PyPDFLoader
from langchain.chat_models import ChatOpenAI
from langchain.chains.summarize import load_summarize_chain
from langchain.chains.mapreduce import MapReduceChain # 引入 MapReduceChain用于处理大文件
from langchain.callbacks import get_openai_callback # 引入OpenAI Callback，用于计算费用
import time #用于计算耗时

pdf_file = "TheEconomist.2024.01.27.pdf"

# 创建 PDF 加载器
loader = PyPDFLoader(pdf_file)

# 加载 PDF 文件的第 21 到第 121 页
# 注意：页码索引从 0 开始，因此第 21 页是索引 20，第 121 页是索引 120
pages_21_to_121 = loader.load()[20:120]

# 创建模板
system_template = "你是北京语言大学高级翻译学院教授高级英语的教师，在给学生讲解经济学人之前会先对文章的核心内容进行分析"
user_template = "请总结以下经济学人文章的核心观点和论据：{text}"

template = PromptTemplate(input_variables=["text"], template=user_template)

refine_template = (
    "你的工作是编写最终摘要\n"
    "我们已经提供了一定程度的现有摘要: {existing_answer}\n"
    "我们有机会完善现有的摘要"
    "(only if needed) 下面有更多背景信息.\n"
    "------------\n"
    "{text}\n"
    "------------\n"
    "鉴于新的背景，完善原始摘要"
    "如果上下文没有用，则返回原始摘要。"
)

refine_prompt = PromptTemplate(
    input_variables=["existing_answer", "text"],
    template=refine_template,
)

# 创建语言模型
llm = ChatOpenAI(temperature=0.7, model_name="gpt-3.5-turbo-1106")

# 创建链

chain = load_summarize_chain(llm,chain_type="refine", question_prompt=template, refine_prompt=refine_prompt)
total_cost = 0  # 初始化总费用
total_time = 0  # 初始化总耗时

# 记录开始时间
start_time = time.time()

# 使用OpenAI Callback计算费用
with get_openai_callback() as cb:
    # 运行语言链
    chain_output = chain.run(pages_21_to_121)

    # 本次调用的费用
    cost = cb.total_cost
    total_cost = cost  # 更新总费用

    # 计算完成这一行的耗时
    row_time = time.time() - start_time
    total_time = row_time  # 更新总耗时

    # 输出翻译结果
    print(f"摘要: {chain_output}\n耗时：{total_time}秒\n耗费：{total_cost}美元\n")

摘要: 根据提供的背景信息，原始摘要已经涵盖了移民潮对美国政治的影响以及拜登总统可能因此失去选举的描述和评价。因此不需要根据新的背景信息来完善摘要。
耗时：502.1684205532074秒
耗费：0.068654美元



使用refine模式得到的摘要为：

摘要: 根据提供的背景信息，原始摘要已经涵盖了移民潮对美国政治的影响以及拜登总统可能因此失去选举的描述和评价。因此不需要根据新的背景信息来完善摘要。

耗时：502.1684205532074秒

耗费：0.068654美元

这个摘要也并没有达到预期的目标。

备注

三种模式的优缺点总结见这篇文章：https://juejin.cn/post/7238110426147504184



**总结**

本课的核心内容是从PDF文件读取内容并进行总结。当PDF文件页数很多时，可以摘取其中的部分内容进行总结，如果需要对长文件进行总结需要考虑使用的模型支持的最大上下文，如果想节约成本，则可以考虑使用langchain的三种基础模式：stuff、map_reduce和refine。